### PR TITLE
Production build script

### PR DESCRIPTION
**Pull request by cdelahousse**
- uses r.js to optimize the client
- uses uglify to minify the bookmarklet.

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -1,3 +1,4 @@
 npm-debug.log
 *.DS_Store
 *~
+production*
```
Comment (spratt) [inline .gitignore]: This is weird.  Aren't we just ignoring production/ and production_tmp/ ?  I think we should ignore those explicitly.  [non-blocking]

File: bin/build_production_all.sh [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+# Configuration                                                               #
+###############################################################################
+
+CLIENT_DIR="$PWD/src/client"
+CLIENT_SCRIPTS_DIR="$CLIENT_DIR/scripts"
+
+BOOKMARKLET_DIR="$PWD/src/bookmarklet"
+
+#Allow production-build directory to be specified by command line args
+TARGET="$PWD/$1"
```
Comment (spratt) [inline bin/build_production_all.sh:16]: This still feels like it should be a Makefile.  [non-blocking]

File: src/client/scripts/app.build.js [hunk 3]
```diff
@@ -1,20 +1,8 @@
...
-    modules: [
-        //Optimize the application files. jQuery is not 
-        //included since it is already in require-jquery.js
-        {
-            name: "main"
-        }
-    ]
+  appDir: '../', //Where the client is located
+  baseUrl: 'scripts/', //Where look for scripts by default
+  dir: '../../../production', //Where to built the optimized project
```
Comment (spratt) [inline src/client/scripts/app.build.js]: Where to build\* [non-blocking]

File: src/client/scripts/app.build.js [hunk 4]
```diff
@@ -1,20 +1,8 @@
...
-
-    modules: [
-        //Optimize the application files. jQuery is not 
-        //included since it is already in require-jquery.js
-        {
-            name: "main"
-        }
-    ]
+  appDir: '../', //Where the client is located
+  baseUrl: 'scripts/', //Where look for scripts by default
```
Comment (spratt) [inline src/client/scripts/app.build.js]: Where to\* look [non-blocking]

File: src/client/scripts/app.build.js [hunk 5]
```diff
@@ -1,20 +1,8 @@
...
-    },
-
-    modules: [
-        //Optimize the application files. jQuery is not 
-        //included since it is already in require-jquery.js
-        {
-            name: "main"
-        }
-    ]
+  appDir: '../', //Where the client is located
```
Comment (spratt) [inline src/client/scripts/app.build.js]: What is the client in this context? [blocking]

File: src/bookmarklet/README.md [hunk 6]
```diff
@@ -14,12 +14,15 @@ move everything to them to where they can be served.
 Run `make build` to build the test-build version of the bookmarklet and vendors
 and move them to where files are are served. 
 
+Run `make build-local` to build the project locally.
```
Comment (spratt) [inline src/bookmarklet/README.md]: This makes it seem like the normal build is remote.  I really don't know what you mean here.  [blocking]

File: bin/build_production_all.sh [hunk 7]
```diff
@@ -0,0 +1,84 @@
...
+if [ -z "$1" ]
+then
```
Comment (spratt) [inline bin/build_production_all.sh]: the convention is to put `then` on the same line as `if`.  Such as in the following example:

```
if [ condition ]; then
```

Not sure if this should be [blocking].  It's a minor issue, but it breaks convention.

File: bin/build_production_all.sh (hunk 2 above)
Comment (cdelahousse) [inline bin/build_production_all.sh:16]: I would rather not

Comment (spratt) [inline bin/build_production_all.sh:16]: Okay, I can rewrite it later.  Any shell file whose entire function is to generate files is better written as a Makefile.

File: bin/build_production_all.sh [hunk 8]
```diff
@@ -0,0 +1,84 @@
...
+# Sanity check                                                                #
+###############################################################################
+
+for DIR in "$CLIENT_DIR" "$BOOKMARKLET_DIR"; do
+  test -d "$DIR"
+  exitIfFailed "Cannot find" `basename "$DIR"` ". Try running this \
+    script from the project root."
+done
+
+type uglifyjs >/dev/null 2>&1
```
Comment (spratt) [inline bin/build_production_all.sh:41]: You can do this simpler with `type uglifyjs &> /dev/null`

**Pull request merged**